### PR TITLE
Clean up search code and unify function returned values

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
This PR is a cleanup: there is no changes in the search results or in the UI.

Depending if it was "literal search" or not, it was either returning booleans or integers. It's pretty bad so instead it all returns integers.

Another thing I did was to move the add and checks into a `addIntoResults` function to simplify things.

Last thing: I removed a loop in the `sortResults` function and moved its code directly into the first loop.

All these changes are done to make #90630 much smaller.

r? @jsha

### CODE REVIEW

**Review by jsha — CHANGES_REQUESTED** [review]

File: src/librustdoc/html/static/js/search.js [hunk 1]
```diff
@@ -363,24 +361,23 @@ window.initSearch = function(rawSearchIndex) {
...
                             for (x = 0; x < len; ++x) {
                                 firstGeneric = val.generics[x];
                                 if (elems[firstGeneric]) {
                                     elems[firstGeneric] -= 1;
                                 } else {
-                                    allFound = false;
-                                    break;
+                                    // Something wasn't found and this is a literal search so
+                                    // abort and return 1.
+                                    return 1;
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:372]: Actually I should return `MAX_LEV_DISTANCE + 1` just like everywhere else.

File: src/librustdoc/html/static/js/search.js [hunk 2]
```diff
@@ -441,19 +439,15 @@ window.initSearch = function(rawSearchIndex) {
                         continue;
                     }
                     tmp = checkType(tmp, val, literalSearch);
-                    if (literalSearch) {
-                        if (tmp) {
-                            return true;
-                        }
+                    if (tmp === 0) {
+                        return 0;
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:443]: Not exactly: before it was returning bool with `literalSearch` was `true` and integer the rest of the time, now it's always returning integers.

However, documentation would definitely be a nice improvement.

File: src/librustdoc/html/static/js/search.js [hunk 3]
```diff
@@ -173,19 +173,17 @@ window.initSearch = function(rawSearchIndex) {
         }
 
         function sortResults(results, isType) {
+            var result;
             var ar = [];
             for (var entry in results) {
                 if (hasOwnPropertyRustdoc(results, entry)) {
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:179]: Good idea. Just like you suggested, I'll do it in another PR.

**Comment by GuillaumeGomez** [comment]
Updated!

File: src/librustdoc/html/static/js/search.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:179]: Actually, we can't use `forEach` on dict/Object... T_T

[the remark this answers is not in the document]
Comment (jsha) [reply src/librustdoc/html/static/js/search.js:179]: We can use [Object.entries](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries).

File: src/librustdoc/html/static/js/search.js [hunk 4]
```diff
@@ -173,19 +173,17 @@ window.initSearch = function(rawSearchIndex) {
         }
 
         function sortResults(results, isType) {
+            var result;
```
Comment (jsha) [inline src/librustdoc/html/static/js/search.js:176]: Move this down to line 180 and combine it with the assignment:

```
var result = results[entry];
```

File: src/librustdoc/html/static/js/search.js [hunk 5]
```diff
@@ -430,6 +428,10 @@ window.initSearch = function(rawSearchIndex) {
             return Math.min(lev_distance, tmp_lev) + 1;
         }
 
+        // This function looks if the object (`obj`) has an argument with the given type (`val`).
+        //
+        // Returns the lev distance of the best match it found (if there are inputs), otherwise
+        // returns MAX_LEV_DISTANCE + 1`.
```
Comment (jsha) [inline src/librustdoc/html/static/js/search.js:434]:
```suggestion
        // This function checks if the object (`obj`) has an argument with the given type (`val`).
        //
        // Returns a Levenshtein distance to the best match. If there is no match, returns MAX_LEV_DISTANCE + 1.
```

File: src/librustdoc/html/static/js/search.js [hunk 6]
```diff
@@ -612,6 +606,34 @@ window.initSearch = function(rawSearchIndex) {
...
+        // following condition:
+        //
+        // * If it is a "literal search" (`isExact`), then `lev` must be 0.
+        // * If it is not a "literal search", `lev` must be <= `MAX_LEV_DISTANCE`.
+        //
+        // `fullId` is the key used of the object we use for the `res` map.
+        //
+        // `id` is the index in both `searchWords` and `searchIndex` arrays for this element.
+        //
+        // `index` is used as a value when sorting results.
```
Comment (jsha) [inline src/librustdoc/html/static/js/search.js:619]: "is used as a value when sorting results" doesn't tell me much. :-) Can you explain more? What type is it? How is it used in sorting?

File: src/librustdoc/html/static/js/search.js [hunk 7]
```diff
@@ -612,6 +606,34 @@ window.initSearch = function(rawSearchIndex) {
             onEach(crateAliases, pushFunc);
         }
 
+        // This function adds the given result into the provided `res` map if it matches the
+        // following condition:
+        //
+        // * If it is a "literal search" (`isExact`), then `lev` must be 0.
+        // * If it is not a "literal search", `lev` must be <= `MAX_LEV_DISTANCE`.
+        //
+        // `fullId` is the key used of the object we use for the `res` map.
```
Comment (jsha) [inline src/librustdoc/html/static/js/search.js:615]: What's the type of `fullId`? What are the expected contents of the `res` map?

**Review by jsha — CHANGES_REQUESTED** [review]
It looks like you forgot to add a function comment on `checkType`?

File: src/librustdoc/html/static/js/search.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:179]: Yes but I read that the performance were less good than a simple `for` loop. To be investigated...

**Comment by GuillaumeGomez** [comment]
Updated! :)

File: src/librustdoc/html/static/js/search.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (jsha) [reply src/librustdoc/html/static/js/search.js:178]: Looks like there was a post in 2018 saying it was slower: https://medium.com/hackernoon/5-techniques-to-iterate-over-javascript-object-entries-and-their-performance-6602dcb708a8

But it had a followup, also in 2018, saying Object.keys() is now faster: https://gists.cwidanage.com/2018/06/how-to-iterate-over-object-entries-in.html

I'm kinda surprised that there would be any significant difference given how similar the operations are and how optimized JS runtimes are.

File: src/librustdoc/html/static/js/search.js [hunk 8]
```diff
@@ -344,7 +341,11 @@ window.initSearch = function(rawSearchIndex) {
             return MAX_LEV_DISTANCE + 1;
         }
 
-        // Check for type name and type generics (if any).
+        // This function checks if the object (`obj`) matches the given type (`val`) and its
+        // generics (if any).
+        //
+        // Returns a Levenshtein distance to the best match. If there is no match, returns
+        // `MAX_LEV_DISTANCE + 1`.
```
Comment (jsha) [inline src/librustdoc/html/static/js/search.js:348]: Can you define the types of `obj`, `val`, and `literalSearch`? In other places in this file, such parameters are documented using [jsdoc](https://jsdoc.app/about-getting-started.html#adding-documentation-comments-to-your-code) syntax, like so:

```js
/**
 * Executes the query and builds an index of results
 * @param  {[Object]} query      [The user query]
 * @param  {[type]} searchWords  [The list of search words to query
 *                                against]
 * @param  {[type]} filterCrates [Crate to search in if defined]
 * @return {[type]}              [A search index of results]
 */
 ```
 
We don't use jsdoc AFAIK, but it's a good syntax for formalizing information about parameters and return types. Also worth noting: I'm not sure why the types are wrapped in brackets. But I think it's okay to follow local style, and we can do a pass later to bring all jsdoc comments in line with the spec.

**Review by jsha — CHANGES_REQUESTED** [review]

File: src/librustdoc/html/static/js/search.js (hunk 8 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:348]: I'm definitely not a big fan of the syntax but why not. :)

Comment (jsha) [reply src/librustdoc/html/static/js/search.js:348]: Fair enough, but I think a syntax that exists and has tooling built for it is much better than no syntax. :-)

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:348]: No sorry, not possible. It's just impossible to read. I'll add the types but the comment is just not human readable...

Comment (jsha) [reply src/librustdoc/html/static/js/search.js:348]: Do you feel that way even for the properly formatted jsdoc style (from https://jsdoc.app/about-getting-started.html#adding-documentation-comments-to-your-code)?

```
/**
 * Represents a book.
 * @constructor
 * @param {string} title - The title of the book.
 * @param {string} author - The author of the book.
 */
function Book(title, author) {
}
```

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:348]: Oh, that's better! That I can do.

**Comment by GuillaumeGomez** [comment]
Updated with jsdoc comments.

File: src/librustdoc/html/static/js/search.js [hunk 9]
```diff
@@ -684,39 +715,23 @@ window.initSearch = function(rawSearchIndex) {
...
                     if (input === "*") {
                         is_module = true;
                     } else {
-                        var allFound = true;
-                        for (it = 0, len = inputs.length; allFound && it < len; it++) {
+                        var allFound = 0;
+                        for (it = 0, len = inputs.length; allFound === 0 && it < len; it++) {
                             allFound = checkType(type, inputs[it], true);
                         }
                         in_args = allFound;
```
Comment (jsha) [inline src/librustdoc/html/static/js/search.js:729]: This loop is pretty confusing to me. I tried to rewrite it with different variable names and a structure that makes its intended meaning clearer (at least I think so!).

```suggestion
                        var firstNonZeroDistance = 0;
                        for (var it = 0, len = inputs.length; it < len; it++) {
                            var distance = checkType(type, inputs[it], true);
                            if (distance > 0) {
                                firstNonZeroDistance = distance;
                                break;
                            }
                        }
                        in_args = firstNonZeroDistance;
```

**Review by jsha — CHANGES_REQUESTED** [review]

**Comment by GuillaumeGomez** [comment]
@jsha Good idea, done as well.

**Comment by camelid** [comment]
> Doesn't have to be part of this PR, but: we are allowed to use ECMASCript 5 features, right?

I think at this point we can even use ES6, but I'm not certain.

**Comment by GuillaumeGomez** [comment]
> I think at this point we can even use ES6, but I'm not certain.

We mentioned it yes. Let's wait for the year to be over before making the change to ensure that we're not missing something.

**Pull request merged**